### ISSUE TITLE
Premake5 trying to include /usr/bin/bitmap instead of folder with premake5 script

### ISSUE THREAD

**Issue by dankmolot** [premake/premake-core#1783]
## What seems to be the problem?
Premake5 `5.0-beta1` tries to include bitmap instead of folder with premake5 script. Here is result:
```
$ premake5 gmake --gmcommon="$HOME/garrysmod_common"
Selected version 3 of the project generator
Error: /usr/bin/bitmap:1: unexpected symbol near '<\127>'
```

## What did you expect to happen?
Expected successful building of configuration.

## What have you tried so far?
Renamed `bitmap` binary, so premake5 can't find it:
```
$ sudo mv /usr/bin/bitmap /usr/bin/bitmap.old
$ premake5 gmake --gmcommon="$HOME/garrysmod_common"
Selected version 3 of the project generator
Building configurations...
Running action 'gmake'...
Generated projects/linux/gmake/Makefile...
Generated projects/linux/gmake/gmsv_network.make...
Done (684ms).
```

Also the older version of premake5 `5.0 alpha 16` works fine.
```
$ ./premake5 gmake --gmcommon="$HOME/garrysmod_common"
Selected version 3 of the project generator
Building configurations...
Running action 'gmake'...
Generated projects/linux/gmake/Makefile...
Generated projects/linux/gmake/gmsv_network.make...
Done (650ms).
```

## How can we reproduce this?
Creating project with bitmap folder
```lua
-- premake5.lua
print('Including bitmap')
include('bitmap')

-- bitmap/premake5.lua
print('Hello from bitmap!')
```

Try build configuration with premake5

`Premake5 5.0-beta1`:
```
$ premake5 gmake
Including bitmap
Error: /usr/bin/bitmap:1: unexpected symbol near '<\127>'
```

`Premake5 5.0 alpha 16`:
```
$ premake5 gmake
Including bitmap
Hello from bitmap!
Building configurations...
Running action 'gmake'...
Done (14ms).
```

## What version of Premake are you using?
Premake 5.0-beta1 from [releases](https://github.com/premake/premake-core/releases)

## Anything else we should know?
Premake5 was running on debian 10

**Comment by dankmolot** [comment]
Interesting fact. The problem is not with the premake5 version, but with the location of the binary. If premake5 is in /usr/bin, it will search /usr/bin first.

**Comment by nickclark2016** [comment]
So it sounds like what you're expecting the behavior to be is for it to search the directory that premake was invoked from first for includes, rather than the directory that the premake executable is located in?

**Comment by dankmolot** [comment]
Exactly! Sorry for confusing description of problem. If something is missing, tell me.

Thanks for the answer!

**Comment by nickclark2016** [comment]
No worries, just trying to make sure I've got a full understanding of the problem.  So if I'm correct, the folder structure is like:

```
/usr
    bin
        premake5
        bitmap

/proj
    premake5.lua
    bitmap
         premake5.lua
```

As a work around, could you try forcing it to `include('bitmap/premake5.lua')` and seeing if that is a temporary workaround.  Per the locating scripts documentation, this is definitely incorrect behavior and should be addressed.

**Comment by dankmolot** [comment]
Made the project as you mentioned above. Here's what I got:

When I used `include('bitmap')`
```
$ premake5
Error: /usr/bin/bitmap:1: unexpected symbol near '<\127>'
```

When I used `include('bitmap/premake5.lua')`
```
$ premake5
Hello from Bitmap!
Type 'premake5 --help' for help
```

**Comment by nickclark2016** [comment]
Awesome. So at least we know that including explicit paths still works, just looks like our implicit path ordering is different than our documentation/spec says.

**Comment by nickclark2016** [comment]
Just following up here, since I completely dropped the ball here. Will try to take a look at this soon.

**Comment by samsinsane** [comment]
@nickclark2016 without digging too deeply into this, I'm not sure there's a valid solution here. Changing the ordering could prevent a user from loading the file they want to load - as seen in #2456 which would prevent a user from ever loading the file they requested if a file matching one of the other patterns exists.

**Comment by mercury233** [comment]
> [@nickclark2016](https://github.com/nickclark2016) without digging too deeply into this, I'm not sure there's a valid solution here. Changing the ordering could prevent a user from loading the file they want to load - as seen in [#2456](https://github.com/premake/premake-core/pull/2456) which would prevent a user from ever loading the file they requested if a file matching one of the other patterns exists.


I have a different perspective on this.

When `include "abc"` is used, Premake currently searches in this order:

1. `abc` (filename)
2. `abc.lua`
3. `abc/premake5.lua`
4. `abc/premake4.lua`

When `include "abc.lua"` is used, it searches:

1. `abc.lua`
2. `abc.lua.lua`
3. `abc.lua/premake5.lua`
4. `abc.lua/premake4.lua`

Due to how filesystems work, cases 1 and 3/4 cannot coexist. Therefore, changing the search order only affects one scenario: when both `abc.lua` and `abc/premake5.lua` exist, and the user writes `include "abc"`.

Importantly, the behavior of automatically appending the `.lua` extension is not documented in the [documentation](https://premake.github.io/docs/globals/include). It is currently undefined behavior. I believe it is reasonable to refine or change undefined behavior during a major version upgrade.

More critically, the `include` command should not be searching system paths at all, should not expect to find premake lua file from system directories. This behavior is also undocumented and problematic. Changing this behavior will fix this issue too. If some users really rely on system path lookup, I suggest introducing a new, explicit function such as `includefromsystem()` instead.

**Comment by mercury233** [comment]
> More critically, the `include` command should not be searching system paths at all, should not expect to find premake lua file from system directories. This behavior is also undocumented and problematic. Changing this behavior will fix this issue too. If some users really rely on system path lookup, I suggest introducing a new, explicit function such as `includefromsystem()` instead.

I noticed this behavior is documented in `Locating-Scripts.md`, but the actual implementation doesn't match the description or the natural expectation: most would assume the search walks through filenames within each directory, not directories for each filename.
